### PR TITLE
Add geospatial process and GIS pip recipe

### PR DESCRIPTION
**Pull request by Zeitsperre**
## Overview

Changes:

* Added a new GeoData process that clips a raster and sends back geospatial data formats
* Added a new recipe for GIS-enabled processes

## Related Issue / Discussion

https://github.com/bird-house/birdy/pull/189
https://github.com/CSHS-CWRA/RavenPy/pull/39

## Additional Information

Links to other issues or sources.

https://automating-gis-processes.github.io/CSC18/lessons/L6/clipping-raster.html

### CODE REVIEW

**Review by cehbrecht — APPROVED** [review]

**Comment by cehbrecht** [comment]
@Zeitsperre caps test is failing ... need to be updated.

File: emu/processes/wps_ultimate_question.py [hunk 1]
```diff
@@ -24,9 +24,11 @@ def __init__(self):
     @staticmethod
     def _handler(request, response):
         import time
+        sleep_delay = .1
         response.update_status('PyWPS Process started.', 0)
+        time.sleep(sleep_delay)
 
-        sleep_delay = .1
+        response.update_status("Contacting the Deep Thought supercomputer.", 10)
```
Comment (cehbrecht) [inline emu/processes/wps_ultimate_question.py:31]: :smile:

File: setup.py [hunk 2]
```diff
@@ -10,14 +10,15 @@
...
-REQUIRES_PYTHON = ">=2.7.0"
+REQUIRES_PYTHON = ">=3.6.0"
 
 about = {}
 with open(os.path.join(here, 'emu', '__version__.py'), 'r') as f:
     exec(f.read(), about)
 
 reqs = [line.strip() for line in open('requirements.txt')]
 dev_reqs = [line.strip() for line in open('requirements_dev.txt')]
+gis_reqs = [line.strip() for line in open("requirements_gis.txt")]
```
Comment (cehbrecht) [inline setup.py:21]: maybe use "extra" instead of "gis" with `requirements_extra.txt`? 

We might add more optional dependencies.

File: emu/processes/wps_ultimate_question.py (hunk 1 above)
Comment (Zeitsperre) [reply emu/processes/wps_ultimate_question.py:31]: :computer: + :earth_americas: = 42

File: emu/processes/wps_geospatial.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+from pywps import Process, ComplexInput, LiteralOutput, BoundingBoxOutput, ComplexOutput
+from pywps import FORMATS
+import logging
+
+LOGGER = logging.getLogger("PYWPS")
+
+__all__ = ["GeoData"]
+
+
+class GeoData(Process):
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply emu/processes/wps_geospatial.py:15]: I can strip down the examples a bit but I wanted to showcase a fairly simple read/write and export operation for users building off Emu. I can strip down the logic a bit but the Fiona/Rasterio API has been standard for years. Perhaps I can make it so that this example isn't tested if the dependencies aren't there?

File: environment.yml [hunk 4]
```diff
@@ -13,6 +13,9 @@ dependencies:
 # opendap support
 - netcdf4
 - xarray
+# geospatial support
+- rasterio
+- fiona
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply environment.yml:18]: For any meaningful GIS operations, they're needed to read and perform operations on GeoJSON/GeoTIFF and Shapefiles. They're in the environment.yml for that reason (for development) and are not installed in a base installation (hence `requirements_extra.txt`)

File: emu/processes/wps_geospatial.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply emu/processes/wps_geospatial.py:15]: I've simplified it quite a bit, but it would be good for testing purposes to have something that showcases inputs and outputs of GeoJSON and GeoTIFF files. I elected to have the example data I made stay under `tests` as we don't necessary want it in the package (again, it's only for testing/development). I hope this is more in line with our goals.

**Comment by cehbrecht** [comment]
@Zeitsperre @huard It is right that the emu process are just simple examples to get started and check the capabilities of a WPS. 

So, you may look for another geo example? Maybe filtering weather stations?
https://github.com/cedadev/goldfinch/blob/main/goldfinch/processes/wps_get_weather_stations.py

I have not checked this implementation. But a list of weather stations with coords is probably available. You can have an area as input (country, bbox, point, ..., ?) and the list of stations "matching" the area is returned as GeoJson. This could then be visualized as a map (leaflet) in a notebook. Maybe including current temperature values?

File: emu/processes/wps_geospatial.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
+"""
+Process using an application/geo+json and 'image/tiff; application=geotiff' input. Used to test UI interactions.
```
Comment (huard) [inline emu/processes/wps_geospatial.py:2]: Please clarify which interactions you are testing exactly. I still don't understand why we'd need to be making geospatial operations inside Emu. 
Testing a BoundingBox Complex Input would be valuable, but this is not high on our priority list at the moment.

File: emu/processes/wps_output_formats.py [hunk 6]
```diff
@@ -3,20 +3,20 @@
...
 import logging
+from pathlib import Path
+
+from pywps import FORMATS, ComplexOutput, Process
+
 LOGGER = logging.getLogger("PYWPS")
 
 # TODO: can be replaced by eggshell function.
-DATA_DIR = os.path.join(os.path.dirname(__file__), '..', 'data')
+DATA_DIR = Path(__file__).parent.parent.joinpath('data')
```
Comment (huard) [inline emu/processes/wps_output_formats.py:14]: Please do not mix unrelated cosmetic fixes with new features.

File: environment.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (huard) [reply environment.yml:18]: Emu is not in the business of being meaningful in terms of processing. You haven't made clear what interactions this tests.

File: tests/test_wps_chomsky.py [hunk 7]
```diff
@@ -7,7 +7,7 @@
 
 def test_wps_chomsky():
     client = client_for(Service(processes=[Chomsky()]))
-    datainputs = "times=10"
+    datainputs = f"times={10}"
```
Comment (huard) [inline tests/test_wps_chomsky.py:10]: ?

File: tests/test_wps_multiple_outputs.py [hunk 8]
```diff
@@ -10,7 +9,7 @@
 @pytest.fixture
 def resp():
     client = client_for(Service(processes=[MultipleOutputs()], cfgfiles=CFG_FILE))
-    datainputs = 'count={}'.format(5)
+    datainputs = f'count={5}'
```
Comment (huard) [inline tests/test_wps_multiple_outputs.py:12]: What's wrong with "count=5" ?

Comment (Zeitsperre) [reply tests/test_wps_multiple_outputs.py:12]: This makes it pop. I think I just got carried away, will change.

File: tests/test_wps_chomsky.py (hunk 7 above)
Comment (Zeitsperre) [reply tests/test_wps_chomsky.py:10]: Idem from `test_wps_multiple_outputs`

File: emu/processes/wps_geospatial.py (hunk 5 above)
Comment (Zeitsperre) [reply emu/processes/wps_geospatial.py:2]: I can remove this now that I have `wps_geodata.py`

File: emu/processes/wps_geodata.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            store_supported=True,
+            status_supported=True,
+        )
+
+    @staticmethod
+    def _handler(request, response):
+        response.update_status("PyWPS Process started.", 0)
+
+        geojson_out = response.outputs['vector']
+        if geojson_out.data_format.mime_type == FORMATS.GEOJSON.mime_type:
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply emu/processes/wps_geodata.py:56]: There seems to be an issue with the mimetype for geojson files, and I'm trying to figure where it's originating from. I'm not receiving the format that I am asking for in my tests for Birdy and I wanted to be certain that it wasn't happening on this end. When I ask for an object that has an obvious converter, when I ask for something `as_obj`, I receive bytes. This was simply to ensure that, yes, these files conform to the respective mimteypes.

File: emu/processes/wps_geodata.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            geojson_out.file = DATA_DIR.joinpath('Olympus_Mons.geojson')
+
+        geotiff_out = response.outputs['raster']
+        if geotiff_out.data_format.mime_type == FORMATS.GEOTIFF.mime_type:
```
[the remark this answers is not in the document]
Comment (Zeitsperre) [reply emu/processes/wps_geodata.py:60]: Idem (above).

File: emu/processes/wps_geodata.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (huard) [reply emu/processes/wps_geodata.py:56]: This is most likely on the converter side. Ie the client cannot associate a converter to the file, so just reads the content as is.

[the remark this answers is not in the document]
Comment (Zeitsperre) [reply emu/processes/wps_geodata.py:56]: I still haven't been able to get the converters to trigger in my PR https://github.com/bird-house/birdy/pull/189. Not sure what I'm missing.

**Comment by huard** [comment]
@cehbrecht The PR also updates PyWPS to 4.4. If you're ok with this I'll go ahead and merge this.

**Comment by cehbrecht** [comment]
@huard Please merge :) I have also a PR prepared which we can use for the job queue testing.
